### PR TITLE
WIP: boats: rotate boat driver, during boat rotation

### PR DESCRIPTION
**Pull request by gang65**
Currently during boat rotation, the boat driver view do not change.
With this simple patch the rotation is made also for driver.

This way is more natural for vehicles (cars, carts, boats, airplanes etc.)

### CODE REVIEW

**Comment by gang65** [comment]
Unfortunately the movement is not smooth as I would like. Do you know how it should be implemented to improve the smoothest?

**Comment by MoNTE48** [comment]
I think there is no better way to do this (so, better not to do this).
This will work even worse for a server with high globalstep. There are no predictions...

**Comment by gang65** [comment]
Maybe it will be possible to add support such rotation in game engine. For example integrate character with the vehicle (boat, cart, car, etc.), and treat it as single object.

**Comment by paramat** [comment]
Unfortunately i doubt this can be made smooth in MTG, it would probably need engine changes to be done clientside somehow.
In multiplayer with additional network lag i expect this would become even less smooth (i cannot test on a network server though).
I have implemented the same view rotation feature as an option in my 'driftcar' mod, and it is also not smooth, even in singleplayer. I only sometimes use the feature because looking around freely is so useful.

Besides smoothness, if this feature is added please make it optional (with a setting added to settingtypes.txt) and disabled by default.
Players usually want, and often need, to look around freely when driving a boat. From my experience with 'driftcar' mod i would find this unbearable if it was non-optional, i think many other players also would. So i disapprove of the PR if it is not optional.
See #731 for some related discussion.

**Comment by paramat** [comment]
Also, the implementation is messy and potentially problematic, this should only need 1 line of code that sets player view to the boat yaw. Maybe:
`driver_objref:set_look_horizontal(self.object:get_yaw())`
?
The current code suggests player view and boat yaw could slowly become de-synced because you are rotating both independently.
The boat yaw should be calculated first, then the player view set to that.

**Comment by paramat** [comment]
Sorry, i think i misunderstood. It seems this allows player view to be freely moved around, and when the boat turns the player view turns by the same amount? Nice idea, some of my previous comments are invlaid then, but i still think this should be optional.

File: mods/boats/init.lua [hunk 1]
```diff
@@ -159,14 +159,18 @@ function boat.on_step(self, dtime)
...
 					self.object:set_yaw(self.object:get_yaw() + dtime * 0.9)
+					driver_objref:set_look_horizontal(driver_objref:get_look_horizontal() + dtime * 0.9)
 				end
 			elseif ctrl.right then
 				if self.v < -0.001 then
 					self.object:set_yaw(self.object:get_yaw() + dtime * 0.9)
+					driver_objref:set_look_horizontal(driver_objref:get_look_horizontal() + dtime * 0.9)
 				else
 					self.object:set_yaw(self.object:get_yaw() - dtime * 0.9)
+					driver_objref:set_look_horizontal(driver_objref:get_look_horizontal() - dtime * 0.9)
```
Comment (paramat) [inline mods/boats/init.lua:173]: All 4 lines are too long. Max 90 columns (with tab size 4).

**Comment by paramat** [comment]
Tested.
Even in singleplayer with no network lag this is too jerky to be acceptable, unfortunately.
Also, when the boat turns, the player view turns slightly slower, so the 2 yaws get out of sync.

The idea and the intended implementation are nice, but due ot the above i unfortunately must disapprove the PR, sorry :-1:

**Comment by paramat** [comment]
Unacceptably jerky even in singelplayer, on a server it will be even worse, and there is no way currently to do this smoothly.
So closing. Might be possible in future with new engine features.
Thanks for the effort and contribution anyway.

**Pull request closed without merging**